### PR TITLE
Fixes #1198 & 1200:  Options - Tablet (Landscape) (Portrait) (Lowfi)

### LINKED ISSUE #1198: Options - Tablet (Landscape) (Lowfi)
Mocks: https://xd.adobe.com/view/d405de00-a871-4f0f-73a0-f8acef30349b-a234/screen/421aa1ca-594a-452e-8367-c990916a5fe6/  

Implement low-fi UI for **Options** tablet landscape mode

**Target PR date**: 26 July 2020
**Target completion date**: 30 July 2020

### LINKED ISSUE #1200: Options - Tablet (Portrait) (Lowfi)
Mocks: https://xd.adobe.com/view/d405de00-a871-4f0f-73a0-f8acef30349b-a234/screen/20631bfd-3756-47a1-9353-ac0a8e9615cd/  

Implement low-fi UI for **Options** tablet portrait mode

**Target PR date**: 26 July 2020
**Target completion date**: 30 July 2020

### PR DESCRIPTION
**Pull request by MohamedMedhat1998**
## Important Note
The newly added test cases in `StoryTextSizeFragmentTest`,`DefaultAudioFragmentTest`, and `AppLanguageFragmentTest` don't pass on robolectric, but when they are copied to espresso, they pass.
## Explanation

- Fixes #1198 
- Fixes #1200 
- Implemented low-fi UI for **Options** for tablet landscape and portrait mode
- Mocks: [portrait](https://xd.adobe.com/view/d405de00-a871-4f0f-73a0-f8acef30349b-a234/screen/20631bfd-3756-47a1-9353-ac0a8e9615cd/) - [landscape](https://xd.adobe.com/view/d405de00-a871-4f0f-73a0-f8acef30349b-a234/screen/421aa1ca-594a-452e-8367-c990916a5fe6/)
## Screenshots
screenshot-٢٠٢٠-٠٨-١٩_١٥ ٠١ ١٢ ٤٢١
screenshot-٢٠٢٠-٠٨-١٩_١٥ ٠١ ٠٧ ٤٥١
screenshot-٢٠٢٠-٠٨-١٩_١٥ ٠١ ٠٣ ١٥٢
screenshot-٢٠٢٠-٠٨-١٩_١٥ ٠٠ ٤٧ ٩٩٦
screenshot-٢٠٢٠-٠٨-١٩_١٥ ٠٠ ٤٣ ٢٧٨
screenshot-٢٠٢٠-٠٨-١٩_١٥ ٠٠ ٣٨ ٤٩١
## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: app/src/main/java/org/oppia/app/options/OptionControlsViewModel.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/options/OptionControlsViewModel.kt]: At `.` at end of sentence.

File: app/src/main/java/org/oppia/app/options/OptionsActivity.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/options/OptionsActivity.kt]: This should not be used directly in other files. Instead pass it down to the viewmodel via fragment/presenter.

File: app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt]: Make this private

**Review by rt4914 — CHANGES_REQUESTED** [review]
Really nice implementation.
I am unable to run test cases but on glance they look correct. Still I will run them in second attempt but meanwhile make sure that your tests pass on Github Actions first.

File: app/src/main/java/org/oppia/app/options/OptionControlsViewModel.kt
Comment (MohamedMedhat1998) [reply app/src/main/java/org/oppia/app/options/OptionControlsViewModel.kt]: Done

File: app/src/main/java/org/oppia/app/options/OptionsActivity.kt
Comment (MohamedMedhat1998) [reply app/src/main/java/org/oppia/app/options/OptionsActivity.kt]: Done

File: app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt
Comment (MohamedMedhat1998) [reply app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt]: Done

**Comment by MohamedMedhat1998** [comment]
> Really nice implementation.
> I am unable to run test cases but on glance they look correct. Still I will run them in second attempt but meanwhile make sure that your tests pass on Github Actions first.

@rt4914 The test cases fail on Robolectric but pass on Espresso (when they are copied there).  
So, the PR itself contains newly added test cases that fail, is it what causing the tests to fail on github actions?

**Comment by rt4914** [comment]
> > Really nice implementation.
> > I am unable to run test cases but on glance they look correct. Still I will run them in second attempt but meanwhile make sure that your tests pass on Github Actions first.
> 
> @rt4914 The test cases fail on Robolectric but pass on Espresso (when they are copied there).
> So, the PR itself contains newly added test cases that fail, is it what causing the tests to fail on github actions?

@MohamedMedhat1998 I was facing similar issue on #1682 and when I merged with latest develop and the test cases are not passing on CI. So can you please update it with latest develop and after that can easily investigate if we still face any issue.

**Comment by MohamedMedhat1998** [comment]
> 
> 
> > > Really nice implementation.
> > > I am unable to run test cases but on glance they look correct. Still I will run them in second attempt but meanwhile make sure that your tests pass on Github Actions first.
> > 
> > 
> > @rt4914 The test cases fail on Robolectric but pass on Espresso (when they are copied there).
> > So, the PR itself contains newly added test cases that fail, is it what causing the tests to fail on github actions?
> 
> @MohamedMedhat1998 I was facing similar issue on #1682 and when I merged with latest develop and the test cases are not passing on CI. So can you please update it with latest develop and after that can easily investigate if we still face any issue.

@rt4914 I have updated with the latest develop.  
Please check the tests in `StoryTextSizeFragmentTest`,`DefaultAudioFragmentTest`, and `AppLanguageFragmentTest` because some of them are failing on robolectric, but when they are moved to espresso, they work.

**Comment by MohamedMedhat1998** [comment]
(Adding Ben) (+24 hours at Rajat's side)
@BenHenning Please check the tests in `StoryTextSizeFragmentTest`,`DefaultAudioFragmentTest`, and `AppLanguageFragmentTest` because some of them are failing on robolectric, but when they are moved to espresso, they work.

File: app/src/test/java/org/oppia/app/testing/options/AppLanguageFragmentTest.kt
Comment (BenHenning) [inline app/src/test/java/org/oppia/app/testing/options/AppLanguageFragmentTest.kt]: Please abbreviate. We shouldn’t disable lint checks.

**Comment by BenHenning** [comment]
@MohamedMedhat1998 I don't have more insight than I did over the weekend currently. You'll need to step debug through these tests on Robolectric & Espresso to better understand what's different about how the code under test is behaving on one versus the other. Once you better understand these differences, I might be able to help you find the root cause.

**Comment by MohamedMedhat1998** [comment]
> @MohamedMedhat1998 I don't have more insight than I did over the weekend currently. You'll need to step debug through these tests on Robolectric & Espresso to better understand what's different about how the code under test is behaving on one versus the other. Once you better understand these differences, I might be able to help you find the root cause.

@rt4914 @BenHenning After debugging, I think that the problem is related to `liveData` and its observers.  
I have searched on the internet and found a similar issue on the Robolectric repository ([issue link](https://github.com/robolectric/robolectric/issues/3234)). There are two solutions in the issue thread:  
1. Use `.handleLifecycleEvent()`, but it is no longer supported
2. Use `shadowOf(getMainLooper()).runOneTask()` but it isn't working with me (tests fail too) (Expected "French" found "No Audio")
```
@Test
  @Config(qualifiers = "sw600dp")
  @LooperMode(LooperMode.Mode.PAUSED)
  fun test() { 
    launch<OptionsActivity>(createOptionActivityIntent(0, true)).use {
      shadowOf(getMainLooper()).runToEndOfTasks()
      onView(
        atPositionOnView(
          R.id.options_recyclerview,
          2,
          R.id.audio_laguage_item_layout
        )
      ).perform(
        click()
      )
      shadowOf(getMainLooper()).runToEndOfTasks()
      onView(
        atPositionOnView(
          R.id.audio_language_recycler_view,
          2,
          R.id.language_radio_button
        )
      ).perform(
        click()
      )
      shadowOf(getMainLooper()).runToEndOfTasks()
      onView(
        atPositionOnView(
          R.id.options_recyclerview,
          2,
          R.id.audio_language_text_view
        )
      ).check(
        matches(withText("French"))
      )
      shadowOf(getMainLooper()).runToEndOfTasks()
    }
  }
```

**Comment by BenHenning** [comment]
Seems potentially related to #322. @MohamedMedhat1998 we need to better understand why it's not being delivered. Unlike that Robolectric thread, Oppia Android has a rich set of test infrastructure now that ensures all background & foreground tasks complete. I've found that in some circumstances in the past, a live data not being notified was either due to a real bug, or a missing runCurrent() call somewhere in the test.

Actually, I didn't look closely enough. Mohamed, have you tried adding ``testCoroutineDispatchers.runCurrent()`` calls after *each* operation that is changing something (e.g. each click(), scroll, etc.)?

**Comment by rt4914** [comment]
> Seems potentially related to #322. @MohamedMedhat1998 we need to better understand why it's not being delivered. Unlike that Robolectric thread, Oppia Android has a rich set of test infrastructure now that ensures all background & foreground tasks complete. I've found that in some circumstances in the past, a live data not being notified was either due to a real bug, or a missing runCurrent() call somewhere in the test.
> 
> Actually, I didn't look closely enough. Mohamed, have you tried adding `testCoroutineDispatchers.runCurrent()` calls after _each_ operation that is changing something (e.g. each click(), scroll, etc.)?

@BenHenning Me and @MohamedMedhat1998 did try ``testCoroutineDispatchers.runCurrent()`` after each action and still the test fails with same reasons as mentioned above.

**Comment by BenHenning** [comment]
Out of curiosity, how have these tests been tested on Espresso? Given that they're in the test/ folder, they're not actually configured to run with instrumentation (and when I try, I get 'No tests were found' as expected for a non-shared test suite).

Edit: also, how did you use testCoroutineDispatchers if DefaultAudioFragmentTest isn't an injected test?

**Comment by BenHenning** [comment]
From my investigation for ``DefaultAudioFragmentTest.testAudioLanguage_clickAudioLanguage_changeAudioLanguageToFrench_checkOptionsFragmentIsUpdatedCorrectly``:
- The click to change the language is changing the audioLanguage in ``OptionsFragmentPresenter`` (by the way, why do we need this variable?) and is initiating an update for the profile. Note that the notify data set changed is wrong here since the adapter list hasn't actually changed yet--that's contingent on the view model's livedata observer getting called back
- UI-thread queued measure calls are binding the default value for audio language, indicating that the update is not actually propagating
- ``processProfileList`` in ``OptionControlsViewModel`` _is_ being called, but with default values (it's probably being called because of the notify data changed call, not because of a change to the profile)
- ProfileManagementController.updateAudioLanguage isn't actually updating the data store, probably because the test never yields the main thread to let the background thread perform its operation
- Actually, it seems like it is calling update, but it's shortcircuiting due to the profile being updated not existing
- It seems like the test isn't actually setting up or logging into any profiles--this is putting the profile system in an invalid state (**action item**: you should add observers to _all_ live datas returned by the profile controller's update methods since these will tell you if they pass/fail--they're clearly failing here. I suggest adding log statements for the cases when they fail.)

I suspect that the Espresso runs of these tests only pass because you already had profiles setup on the local device. Robolectric is more correct here: it's always running the tests in a clean state, uncovering the test setup issue. Restructuring your test like so fixes it for me:

https://gist.github.com/BenHenning/0582c3bc1f8cb03917b030dc8238f075/revisions

**Comment by rt4914** [comment]
> * OptionsFragmentPresenter

```
      getOptionControlsItemViewModel().getAppLanguage(AppLanguage.ENGLISH_APP_LANGUAGE) -> {
        profileManagementController.updateAppLanguage(
          profileId,
          AppLanguage.ENGLISH_APP_LANGUAGE
        ).observe(fragment, Observer { appLanguageResult ->
          if (appLanguageResult.isSuccess()) {
            appLanguage = AppLanguage.ENGLISH_APP_LANGUAGE
          }
        })
      }
```

**Comment by rt4914** [comment]
> getOptionControlsItemViewModel().getAppLanguage(AppLanguage.ENGLISH_APP_LANGUAGE) -> {
>         profileManagementController.updateAppLanguage(
>           profileId,
>           AppLanguage.ENGLISH_APP_LANGUAGE
>         ).observe(fragment, Observer { appLanguageResult ->
>           if (appLanguageResult.isSuccess()) {
>             appLanguage = AppLanguage.ENGLISH_APP_LANGUAGE
>           }
>         })
>       }

@MohamedMedhat1998 Next steps / commits:
1. Update test cases to use `initialiseProfiles`
2. Resolve merge conflicts.
3. Update implementation to observe results.
4. You might need to Recheck test cases after step 3.

**Comment by MohamedMedhat1998** [comment]
> Out of curiosity, how have these tests been tested on Espresso? Given that they're in the test/ folder, they're not actually configured to run with instrumentation (and when I try, I get 'No tests were found' as expected for a non-shared test suite).

I just copied the tests from Robolectric files to Espresso files :D

  


> Edit: also, how did you use testCoroutineDispatchers if DefaultAudioFragmentTest isn't an injected test?

I didn't use `testCoroutineDispatchers` while I was trying on Espresso

**Comment by MohamedMedhat1998** [comment]
>     * The click to change the language is changing the audioLanguage in `OptionsFragmentPresenter` (by the way, why do we need this variable?) and is initiating an update for the profile

Because when we are on a tablet device, the `AudioLanguageActivity` isn't started, so we update the language using the `OptionsFragmentPresenter` instead of `AudioLanguageActivityPresenter`  


>     * Note that the notify data set changed is wrong here since the adapter list hasn't actually changed yet--that's contingent on the view model's livedata observer getting called back

I will fix that using Rajat snippet in his comment.

>     * It seems like the test isn't actually setting up or logging into any profiles--this is putting the profile system in an invalid state (**action item**: you should add observers to _all_ live datas returned by the profile controller's update methods since these will tell you if they pass/fail--they're clearly failing here. I suggest adding log statements for the cases when they fail.)

I think this will be solved also using the code Rajat added in his comment.  


> I suspect that the Espresso runs of these tests only pass because you already had profiles setup on the local device. Robolectric is more correct here: it's always running the tests in a clean state, uncovering the test setup issue. Restructuring your test like so fixes it for me:
> 
> https://gist.github.com/BenHenning/0582c3bc1f8cb03917b030dc8238f075/revisions

Thanks a lot, Ben. This really helped.

File: app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt:161]: I suggest also logging here & elsewhere when it fails since that can help with investigations in situations like the one you were in before with the failing tests.

File: app/src/test/java/org/oppia/app/testing/options/AppLanguageFragmentTest.kt
Comment (MohamedMedhat1998) [reply app/src/test/java/org/oppia/app/testing/options/AppLanguageFragmentTest.kt]: Done

File: app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt
Comment (MohamedMedhat1998) [reply app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt:161]: Done

Comment (rt4914) [reply app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt:161]: Resolving this on behalf of Ben and also to unblock Mohamed.

File: app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt]: You cannot have the same logger everywhere, the error should be different for each logger otherwise if anyone faces this error how would they know which error is related to which function?

              consoleLogger.e(
                READING_TEXT_SIZE_TAG,
                READING_TEXT_SIZE_ERROR + ": small text size",
                it.getErrorOrNull()
              )

File: app/src/test/java/org/oppia/app/testing/options/AppLanguageFragmentTest.kt
Comment (rt4914) [reply app/src/test/java/org/oppia/app/testing/options/AppLanguageFragmentTest.kt]: Resolving this to unblock Mohamed.

File: app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt
Comment (rt4914) [reply app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt:161]: Resolving this to unblock Mohamed.

**Review by rt4914 — CHANGES_REQUESTED** [review]
Nit change suggested in logger.

File: app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt
Comment (MohamedMedhat1998) [reply app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt]: Done

File: app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/app/options/OptionsFragmentPresenter.kt]: You can write this directly like this:
`$READING_TEXT_SIZE_ERROR: small text size`

Here and everywhere else.

Once done, you can resolve this comment on your own.

**Review by rt4914 — APPROVED** [review]
LGTM, nit change suggested.

**Review by BenHenning — APPROVED** [review]
Approval for codeowners for the Bazel changes (didn't look at the rest of the PR in depth).

**Pull request merged**
